src/parser/func_type.cpp

```diff
@@ -1052,33 +1131,101 @@ Func::Node::apply_signature (const func_signature &signature)
   return true;
 }
 
-void
+bool
 Func::Node::check_arg_compat (const pt_arg_type &arg_signature, const PT_NODE *arg_node,
-			      argument_compatibility &compat)
+			      signature_compatibility &compat, argument_resolve &arg_res)
 {
-  compat.m_type = PT_TYPE_NONE;
-  compat.m_compat = type_compatibility::INCOMPATIBLE;
+  arg_res.m_type = PT_TYPE_NONE;
 
   // todo - equivalent type & coercible type checks should all be in a the same place to have a better view of how
   //        each type can convert to another
 
   if (cmp_types_equivalent (arg_signature, arg_node->type_enum))
     {
-      compat.m_compat = type_compatibility::EQUIVALENT;
-      compat.m_type = pt_get_equivalent_type (arg_signature, arg_node->type_enum);
+      arg_res.m_type = pt_get_equivalent_type (arg_signature, arg_node->type_enum);
     }
   else if (cmp_types_castable (arg_signature, arg_node->type_enum))
     {
       compat.m_compat = type_compatibility::COERCIBLE;
-      compat.m_type = pt_get_equivalent_type (arg_signature, arg_node->type_enum);
+      arg_res.m_type = pt_get_equivalent_type (arg_signature, arg_node->type_enum);
+    }
+  else
+    {
+      compat.m_compat = type_compatibility::INCOMPATIBLE;
+      invalid_arg_error (arg_signature, arg_node, *compat.m_signature);
+      return false;
     }
 
   // if compatible, pt_get_equivalent_type should return a valid type. but we need to double-check
-  if (compat.m_compat != type_compatibility::INCOMPATIBLE && compat.m_type == PT_TYPE_NONE)
+  if (arg_res.m_type == PT_TYPE_NONE)
     {
       assert (false);
       compat.m_compat = type_compatibility::INCOMPATIBLE;
+      invalid_arg_error (arg_signature, arg_node, *compat.m_signature);
+      return false;
     }
+
+  // three conditions should be met to require collation inference:
+  //
+  //  1. argument signature should allow collation. e.g. all generic strings allow collations, but json docs and values
+  //     don't allow
+  //
+  //  2. equivalent type should have collation.
+  //
+  //  3. original argument type should have collation. if it doesn't have, it doesn't affect common collation.
+  //     NOTE - if first two conditions are true and this is false, we don't do collation inference, but argument will
+  //            be coerced to common collation.
+  //
+  // todo - what happens when all arguments are maybe??
```

Comment:
@razvanradulescu 

I am not sure how to handle the case with only maybe arguments. Do you think current code would work?

Reply:
The handling of ' TP_DOMAIN_COLL_ACTION collation_flag' is missing from new code.
However, a generic rule is hard to detect from the old code which was handling particular cases individually (INSERT function, GROUP_CONCAT with host variable arguments)